File: src/main/kotlin/magicbook/gtlitecore/integration/justenoughitems/category/SpacePumpRecipeCategory.java
package magicbook.gtlitecore.integration.justenoughitems.category;

import gregtech.api.gui.GuiTextures;
import gregtech.api.util.GTStringUtils;
import gregtech.integration.jei.basic.BasicRecipeCategory;
import magicbook.gtlitecore.api.utils.GTLiteValues;
import magicbook.gtlitecore.integration.justenoughitems.info.SpacePumpRecipeWrapper;
import mcp.MethodsReturnNonnullByDefault;
import mezz.jei.api.IGuiHelper;
import mezz.jei.api.gui.IDrawable;
import mezz.jei.api.gui.IGuiFluidStackGroup;
import mezz.jei.api.gui.IRecipeLayout;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.IRecipeWrapper;
import net.minecraft.client.Minecraft;
import net.minecraft.client.resources.I18n;
import net.minecraftforge.fluids.FluidStack;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.Collections;
import java.util.List;

@MethodsReturnNonnullByDefault
@ParametersAreNonnullByDefault
public final class SpacePumpRecipeCategory extends BasicRecipeCategory<SpacePumpRecipeWrapper, SpacePumpRecipeWrapper>
{

    private static final int SLOT_CENTER = 79;
    private static final int TEXT_START_X = 5;
    private static final int START_POS_Y = 40;

    private int planetId;
    private int fluidId;
    private FluidStack fluid;
    private int fluidAmount;

    private final IDrawable slot;

    private int planetIdLength;
    private int fluidIdLength;
    private int fluidAmountLength;

    public SpacePumpRecipeCategory(IGuiHelper guiHelper)
    {
        super("space_pump_module", "gtlitecore.jei.space_pump_module.name",
                guiHelper.createBlankDrawable(176, 166), guiHelper);
        this.slot = guiHelper.drawableBuilder(GuiTextures.SLOT.imageLocation, 0, 0, 18, 18)
                .setTextureSize(18, 18)
                .build();
    }

    @Override
    public void setRecipe(IRecipeLayout recipeLayout,
                          SpacePumpRecipeWrapper recipeWrapper,
                          IIngredients ingredients)
    {
        IGuiFluidStackGroup fluidStackGroup = recipeLayout.getFluidStacks();
        fluidStackGroup.init(0, true, SLOT_CENTER,
                19, 16, 16,
                1, false, null);
        fluidStackGroup.set(ingredients);

        this.planetId = recipeWrapper.getPlanetId();
        this.fluidId = recipeWrapper.getFluidId();
        this.fluid = recipeWrapper.getFluid();
        this.fluidAmount = recipeWrapper.getFluid().amount;
    }

    @Override
    public void drawExtras(Minecraft mc)
    {
        GTStringUtils.drawCenteredStringWithCutoff(fluid.getLocalizedName(), mc.fontRenderer, 176);
        this.slot.draw(mc, SLOT_CENTER - 1, 18);

        String planetIdName = I18n.format("gtlitecore.jei.space_pump_module.planet_id", this.planetId);
        this.planetIdLength = mc.fontRenderer.getStringWidth(planetIdName);
        mc.fontRenderer.drawString(planetIdName, TEXT_START_X, START_POS_Y, 0x111111);

        String fluidIdName = I18n.format("gtlitecore.jei.space_pump_module.fluid_id", this.fluidId);
        this.fluidIdLength = mc.fontRenderer.getStringWidth(fluidIdName);
        mc.fontRenderer.drawString(fluidIdName, TEXT_START_X, START_POS_Y + FONT_HEIGHT + 1, 0x111111);

        String fluidAmountName = I18n.format("gtlitecore.jei.space_pump_module.fluid_amount", this.fluidAmount);
        this.fluidAmountLength = mc.fontRenderer.getStringWidth(fluidAmountName);
        mc.fontRenderer.drawString(fluidAmountName, TEXT_START_X, START_POS_Y + FONT_HEIGHT * 2 + 1, 0x111111);
    }

    @Override
    public List<String> getTooltipStrings(int mouseX, int mouseY)
    {
        return Collections.emptyList();
    }

    @Override
    public IRecipeWrapper getRecipeWrapper(SpacePumpRecipeWrapper recipeWrapper)
    {
        return recipeWrapper;
    }

    @Override
    public String getModName()
    {
        return GTLiteValues.MODID;
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/integration/justenoughitems/category/SpacePumpRecipeCategory.kt
package gregtechlite.gtlitecore.integration.justenoughitems.category

import gregtech.api.gui.GuiTextures
import gregtech.api.util.GTStringUtils.drawCenteredStringWithCutoff
import gregtech.integration.jei.basic.BasicRecipeCategory
import gregtechlite.gtlitecore.api.MOD_ID
import gregtechlite.gtlitecore.integration.justenoughitems.info.SpacePumpRecipeWrapper
import mezz.jei.api.IGuiHelper
import mezz.jei.api.gui.IRecipeLayout
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.recipe.IRecipeWrapper
import net.minecraft.client.Minecraft
import net.minecraft.client.resources.I18n
import net.minecraftforge.fluids.FluidStack

class SpacePumpRecipeCategory(guiHelper: IGuiHelper) : BasicRecipeCategory<SpacePumpRecipeWrapper, SpacePumpRecipeWrapper>(
    "space_pump_module", "gtlitecore.jei.space_pump_module.name",
    guiHelper.createBlankDrawable(176, 166 - 36), guiHelper)
{

    companion object
    {

        private const val SLOT_CENTER = 79

        @JvmField
        val UID = String.format("%s.space_pump_module", MOD_ID)

    }

    private val slot = guiHelper.drawableBuilder(GuiTextures.SLOT.imageLocation, 0, 0, 18, 18)
        .setTextureSize(18, 18)
        .build()

    override fun setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: SpacePumpRecipeWrapper, ingredients: IIngredients)
    {
        val fluidStackGroup = recipeLayout.fluidStacks
        fluidStackGroup.init(0, true, SLOT_CENTER, 19,
            16, 16, 1, false, null)
        fluidStackGroup.set(ingredients)
    }

    override fun drawExtras(mc: Minecraft)
    {
        slot.draw(mc, SLOT_CENTER - 1, 18)
    }

    override fun getTooltipStrings(mouseX: Int, mouseY: Int): MutableList<String?> = mutableListOf()

    override fun getRecipeWrapper(recipeWrapper: SpacePumpRecipeWrapper): IRecipeWrapper = recipeWrapper

    override fun getModName() = MOD_ID

}
